Combat GUI

Wombatlord opened the pull request:
- Created a button for advancing the turn actions.
- Attached handler
- Composed CommandBarSection for BattleView and disabled drawing turn prompt.
- Button click is equivalent to space key press to advance, both available.

Wombatlord committed:
Combat GUI

- Created a button for advancing the turn actions.
- Attached handler
- Composed CommandBarSection for BattleView and disabled drawing turn prompt.
- Button click is equivalent to space key press to advance, both available.

Wombatlord committed:
CombatInputMode.on_key_press(space) now invokes the on_click handler of the end turn button.

Wombatlord committed:
NinePatchTextures for buttons! Extended arcade.NinePatchTexture for now to override pixelated flag on draw_sized().

Wombatlord committed:
Increased font sizes on buttons and adjusted colours

Wombatlord committed:
Merge branch 'main' of github.com:Wombatlord/adventure-league into battle-buttons

tjweldon commented on src/gui/buttons.py:
```diff
@@ -23,6 +25,45 @@ def command_bar(self) -> list[UIFlatButton]:
+
+        self._atlas.use_uv_texture(0)
+        self._atlas.texture.use(1)
+        self._geometry.render(self._program, vertices=1)
```
If you ever find yourself doing the exact same thing as one of the base classes, this is the syntax to just defer the call up the inheritance hierarchy:
```suggestion
        super().draw_sized(position=position, size=size, pixelated=pixelated, **kwargs)

```

tjweldon commented on src/gui/buttons.py:
```diff
@@ -42,17 +83,32 @@ def _handle(event: UIEvent):
+        UITextureButton: A button with a text label and an attached click handler.
     """
-    btn = UIFlatButton(text=text)
+    main_tex = PixelatedNinePatch(
+        left=1, right=1, bottom=3, top=1, texture=WindowData.buttons[7]
+    )
+    pressed_tex = PixelatedNinePatch(
+        left=1, right=1, bottom=3, top=3, texture=WindowData.buttons[9]
+    )
+    hovered_tex = PixelatedNinePatch(
+        left=1, right=1, bottom=3, top=1, texture=WindowData.buttons[11]
+    )
```
These should probably be supplied as arguments since they are implementation specific

tjweldon replied:
Or use-case specific rather

tjweldon replied:
Actually now that I've looked at more of the code, this should be defined as a dict at the module level, see example on the `end_turn_button` function

tjweldon commented on src/gui/buttons.py:
```diff
@@ -62,8 +118,64 @@ def get_new_missions_handler(event: UIEvent) -> UIEventHandler:
-def get_new_missions_button() -> UIFlatButton:
-    btn = UIFlatButton(text="New Missions")
+def get_new_missions_button() -> UITextureButton:
+    main_tex = PixelatedNinePatch(
+        left=1, right=1, bottom=3, top=1, texture=WindowData.buttons[7]
+    )
+    pressed_tex = PixelatedNinePatch(
+        left=1, right=1, bottom=3, top=3, texture=WindowData.buttons[9]
+    )
+    hovered_tex = PixelatedNinePatch(
+        left=1, right=1, bottom=3, top=1, texture=WindowData.buttons[11]
```
I think WindowData needs to be split into a couple of modules, one that gives access to textures (texture_loading or something), so window_data.py only contains data about the window. If we were to name it accurately now it would be called `window_data_and_sprite_textures.py`, that `and`  in the naming is a signal that it's time to separate some concerns!

tjweldon replied:
Maybe we have a `TextureData` class that handles access to textures that are loaded once on the first module import and then are just returned from the in-memory object. Texture loading with the class as currently defined happens every time we import the window_data module (could be wrong about that, worth verifying).

tjweldon replied:
Even further reading has made me to think that the specific `PixelatedNinePatch.__init__`  arguments are essentially configuration, and we can just instantiate them once in a module with an appropriate name and use the instances wherever we need them.

tjweldon replied:
Even if arcade is doing some sneaky cache magic under the hood to prevent these re-instantiations of the same textures (which it almost certainly is), it removes a bunch of repeated code

tjweldon commented on src/gui/buttons.py:
```diff
@@ -62,8 +118,64 @@ def get_new_missions_handler(event: UIEvent) -> UIEventHandler:
     return btn
+
+
+def end_turn_handler(view) -> UIEventHandler:
```
This function is not the handler, the function it returns (`_handle`) is the handler, so this is a handler getter!

```suggestion
def get_end_turn_handler(view) -> UIEventHandler:

```

tjweldon commented on src/gui/buttons.py:
```diff
@@ -62,8 +118,64 @@ def get_new_missions_handler(event: UIEvent) -> UIEventHandler:
+            view.item_menu_mode_allowed = True
+
+    return _handle
+
+
+def end_turn_button(view) -> UITextureButton:
+    main_tex = PixelatedNinePatch(
+        left=1, right=1, bottom=3, top=1, texture=WindowData.buttons[7]
+    )
+    pressed_tex = PixelatedNinePatch(
+        left=1, right=1, bottom=3, top=3, texture=WindowData.buttons[9]
+    )
+    hovered_tex = PixelatedNinePatch(
+        left=1, right=1, bottom=3, top=1, texture=WindowData.buttons[11]
+    )
+
+    btn = UITextureButton(
+        text="CLICK ME TO ADVANCE!",
+        texture=main_tex,
+        texture_pressed=pressed_tex,
+        texture_hovered=hovered_tex,
+    )
+
+    btn.on_click = end_turn_handler(view)
+
+    return btn
```
```suggestion

button_textures = {
    "texture": PixelatedNinePatch(
        left=1, right=1, bottom=3, top=1, texture=WindowData.buttons[7]
    )
    "texture_pressed": PixelatedNinePatch(
        left=1, right=1, bottom=3, top=3, texture=WindowData.buttons[9]
    )
    "texture_hovered": PixelatedNinePatch(
        left=1, right=1, bottom=3, top=1, texture=WindowData.buttons[11]
    )
}


def end_turn_button(view) -> UITextureButton:
    btn = UITextureButton(
        text="CLICK ME TO ADVANCE!",
        **button_textures
    )

    btn.on_click = end_turn_handler(view)

    return btn

```

tjweldon replied:
^ This will allow you to share texture configuration amongst your button factory functions

tjweldon commented on src/gui/views.py:
```diff
@@ -325,27 +331,55 @@ def switch_to_roster_and_team_panes(self, event: UIEvent = None):
         """
-        btn = UIFlatButton(
-            text="Recruit "
+        main_tex = PixelatedNinePatch(
+            left=1, right=1, bottom=3, top=1, texture=WindowData.buttons[7]
+        )
+        pressed_tex = PixelatedNinePatch(
+            left=1, right=1, bottom=3, top=3, texture=WindowData.buttons[9]
+        )
+        hovered_tex = PixelatedNinePatch(
+            left=1, right=1, bottom=3, top=1, texture=WindowData.buttons[11]
+        )
```
Seems like we might be in need of a module to allow this to be defined once and then shared amongst various places.

tjweldon commented on src/gui/window_data.py:
```diff
@@ -42,6 +42,15 @@ class WindowData:
         margin=1,
     )
 
+    buttons = arcade.load_spritesheet(
+        "./assets/sprites/buttons.png",
+        sprite_height=16,
+        sprite_width=48,
+        columns=6,
+        count=24,
+        margin=0,
+    )
+
```
I think we have strained the `WindowData` abstraction to breaking point. The sooner we break it up, the less painful it will be

Wombatlord committed:
Lazy texture loading implementation for sprite sheets and single texture files.

Wombatlord committed:
formatting

Wombatlord committed:
Better name for TextureSpec type alias

Wombatlord committed:
NinePatchConfig

- Extracted roster and recruit button implementations to buttons.py
- Created Textures package and extracted custom NinePatch impl and texture_data.py
- Created function for getting a configured tuple of PixelatedNinePatches for TextureButtons
- Expressed PixelatedNinePatch parameters as config.

Wombatlord committed:
Extraction

- created button_textures.py to hold factory for button textures created from button texture config.
- corrected incorrect texture assignment

Wombatlord committed:
Better config and loading

- Moved texture_config.py out of global config package and into textures package.
- Added loader functions for creating UITextureButtons from supplied config.
- Removed button_textures.py, replaced with loaders in texture_config.py.
- Escape key exits app from GuildView.

Wombatlord committed:
Better KeyError message to indicate misconfiguration of TextureButton textures
